defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+package defrag
+
+import (
+	"container/list"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"code.google.com/p/gopacket"
+	"code.google.com/p/gopacket/layers"
+)
+
+// Quick and Easy yo use debug code to trace
+// how defrag works.
+var debug debugging = false // or flip to false
+type debugging bool
+
+func (d debugging) Printf(format string, args ...interface{}) {
+	if d {
+		log.Printf(format, args...)
+	}
+}
+
+type Defragmenter interface {
+	// DefragIPv4 takes in an IPv4 packet with a fragment payload.
+	//
+	// It modifies the IPv4 layer in place, returning true if the
+	// modified layer is now a full IPv4 payload.
+	//
+	// If the passed-in IP layer is NOT fragmented, it will
+	// immediately return true without modifying the layer.
+	//
+	// If the IPv4 layer is a fragment and we don't have all
+	// fragments, it will return false and store whatever internal
+	// information it needs to eventually defrag the packet.
+	//
+	// If the IPv4 layer is the last fragment needed to reconstruct
+	// the packet, *ip will be set to the entire defragmented packet,
+	// and the function will return true.
+	DefragIPv4(IPv4 *layers.IPv4) (bool, error)
```

Comment:
I am not sure to understand your point correctly. Would you please give the caller skeleton ?
I mean, if we return "false", the caller already have the packet because he just passed it... there is no new allocation.